components/locale/macros/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+
+fn get_value_from_token_stream(input: TokenStream) -> String {
+    let val = format!("{}", input);
+    if !val.starts_with('"') || !val.ends_with('"') {
+        panic!("Argument must be a string literal.");
+    }
+    let len = val.len();
+    (&val[1..len - 1]).to_string()
+}
+
+fn language_output(input: Option<u64>) -> String {
+    let val = if let Some(raw) = input {
+        format!("Some({})", raw)
+    } else {
+        "None".to_string()
+    };
+    format!("unsafe {{ icu_locale::subtags::Language::from_raw_unchecked({}) }}", val)
+}
+
+fn script_output(input: u32) -> String {
+    format!("unsafe {{ icu_locale::subtags::Script::from_raw_unchecked({}) }}", input)
+}
+
+fn region_output(input: u32) -> String {
+    format!("unsafe {{ icu_locale::subtags::Region::from_raw_unchecked({}) }}", input)
+}
+
+fn variant_output(input: u64) -> String {
+    format!("unsafe {{ icu_locale::subtags::Variant::from_raw_unchecked({}) }}", input)
+}
+
+fn variants_output(input: Option<Box<[icu_locale::subtags::Variant]>>) -> String {
+    if let Some(variants) = input {
+        let v: Vec<_> = variants
+            .iter()
+            .map(|v| {
+                let variant: u64 = v.into_raw();
+                format!("unsafe {{ icu_locale::subtags::Variant::from_raw_unchecked({}) }}", variant)
+            })
+            .collect();
+        let variants = format!("Some(Box::new([{}]))", v.join(", "));
+        format!("icu_locale::subtags::Variants::from_raw_unchecked({})", variants)
+    } else {
+        format!("icu_locale::subtags::Variants::from_raw_unchecked(None)")
+    }
+}
+
+/// A macro allowing for compile-time construction of valid [`Language`] subtag.
+///
+/// The macro will perform canonicalization of the tag.
```

Comment:
Here and elsewhere (this might be a pre-existing issue): I would avoid using the term "perform canonicalization".  UTS 35 defines two types of canonicalization:

- *canonical syntax*
- *canonical form*

You are doing the first but not the second (#218).

Alternative language options:

- The macro will return the tag in canonical syntax.
- The macro will transform the tag into canonical syntax.
- The macro will perform syntax canonicalization of the tag.  (Not as precise, because "syntax canonicalization" is not a term that you can find in UTS 35.)

http://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers

Reply:
Zibi's reply: https://github.com/unicode-org/icu4x/pull/220#issuecomment-695356372